### PR TITLE
Collision tester

### PR DESCRIPTION
**Pull request by Cyan4973**
Add a new test tool, the brute-force collision tester,
used to verify the quality of 64-bit hash functions
with regards to collision ratio.

This kind of analysis is out of `SMHasher` range, this it takes many billions of hashes to observe just a few collisions.

The tool requires a huge amount of memory, to store hash values, or even just to filter candidates.
See `README.md` for details.

### CODE REVIEW

**Comment by Cyan4973** [comment]
That one is strange :
The test error signals that the multi-treading code of the new tester
doesn't compile on `mingw32`.

But it compiles fine on `mingw64`.

Also : the same code is in use within `zstd`, and it compiles fine with both `mingw64` and `mingw32`.

...

File: README.md [hunk 1]
```diff
@@ -62,17 +62,27 @@ thanks to [Takayuki Matsuoka](https://github.com/t-mat) contributions.
...
+
+The new algorithm is much faster than its predecessors,
+for both long and small inputs,
+as can be observed in following graphs :
+
+![XXH3, bargraph](https://user-images.githubusercontent.com/750081/61976096-b3a35f00-af9f-11e9-8229-e0afc506c6ec.png)
+
+![XXH3, latency, random size](https://user-images.githubusercontent.com/750081/61976089-aedeab00-af9f-11e9-9239-e5375d6c080f.png)
+
+To access these new prototypes, one needs to unlock their declaration, using build macro `XXH_STATIC_LINKING_ONLY`.
```
Comment (easyaspi314) [inline README.md]: ...their declaration using the build macro...

File: README.md [hunk 2]
```diff
@@ -62,17 +62,27 @@ thanks to [Takayuki Matsuoka](https://github.com/t-mat) contributions.
...
-    cd vcpkg
-    ./bootstrap-vcpkg.sh
-    ./vcpkg integrate install
-    ./vcpkg install xxhash
+Starting with `v0.7.0`, the library includes a new algorithm, named `XXH3`,
+able to generate 64 and 128-bits hashes.
+
+The new algorithm is much faster than its predecessors,
+for both long and small inputs,
+as can be observed in following graphs :
```
Comment (easyaspi314) [inline README.md]:
```
Which can be observed in the following graphs:
```

File: tests/collisions/README.md [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+which will measure a 64-bit hash algorithm's collision rate
+by generating billions of hashes,
+and comparing the result to an "ideal" target.
+
+The test requires a very large amount of memory.
+By default, it will generate 24 billion of 64-bit hashes,
+requiring __192 GB of RAM__ for their storage.
+The number of hashes can be modified using command `--nbh=`.
+be aware that testing the collision ratio of 64-bit hashes
+requires a very large amount of hashes (several billions) for meaningful measurements.
```
Comment (easyaspi314) [inline tests/collisions/README.md]: Be aware that testing the collision ratio of 64-bit hashes requires a very large amount of hashes (several billion) for meaningful measurements.

File: tests/collisions/README.md [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+#### Build modifier
+
+- `SLAB5` : use alternative pattern generator, more friendly to weak algorithms
+
+#### How to integrate any hash in the tester
+
+The build script is expecting to compile files in `./allcodecs`.
+Put the source code here.
+This also works if the hash is a single `*.h` files.
```
Comment (easyaspi314) [inline tests/collisions/README.md]: file

File: tests/collisions/main.c [hunk 5]
```diff
@@ -0,0 +1,1096 @@
...
+
+    ((unsigned char*)bf)[pos] |= (unsigned char)(nextValue[existingCandidates] << shift);
+    return addCandidates[existingCandidates];
+}
+
+/* Check if provided 64-bit hash is a collision candidate
+ * Requires the slot to be occupied by at least 2 candidates.
+ * return >0 if hash is collision candidate
+ *         0 otherwise (slot unoccupied, or only one candidate)
+ * note: slot unoccupied should not happen in this algorithm,
```
Comment (easyaspi314) [inline tests/collisions/main.c]: if hash is **a** collision candidate

Note: unoccupied slots

File: tests/collisions/main.c [hunk 6]
```diff
@@ -0,0 +1,1096 @@
...
+    if (result) *stringPtr += comSize;
+    return result;
+}
+
+
+#include "pool.h"
+
+/* As some hashes use different algorithms depending on input size,
+ * it can be necessary to test multiple input sizes
+ * to paint an accurage picture on collision performance */
```
Comment (easyaspi314) [inline tests/collisions/main.c]: accurate

**Comment by Cyan4973** [comment]
> Is the space before the punctuation a French thing? I haven't seen it used like that in English.

I'm not sure if it's French, but I was used to it before coming to US.
Since then, I'm regularly reminded that this is not a English thing ...

**Comment by easyaspi314** [comment]
Yeah, I just looked it up, and it *does* appear to be how punctuation works in French. Very interesting.

That explains it quite well.

**Comment by easyaspi314** [comment]
Just leave it as-is. I can fix it this week.

**Pull request merged**